lib/wallaroo_labs/logging/logging-lib.c

```diff
@@ -0,0 +1,391 @@
+ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#include <ctype.h>
+#include <stdarg.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/time.h>
+
+#define ENV_BUF_SIZE  1024
+#define FMT_BUF_SIZE  1024
+#define MAX_SEVERITY  8
+#define MAX_CATEGORY  64
+
+static int _labels_initialized = 0;
+
+static char *_severity_labels[MAX_SEVERITY+1];
+static char *_category_labels[MAX_CATEGORY+1];
+
+static int _cat2sev_threshold[MAX_CATEGORY+1];
+
+/*****************************/
+/* Internal static functions */
+/*****************************/
+
+static int _w_now(char *dst, int dst_size)
+{
+  struct timeval tv;
+
+  gettimeofday(&tv, NULL);
+#ifdef  __APPLE__  
+  return snprintf(dst, dst_size, "%ld.%06d", tv.tv_sec, tv.tv_usec);
+#else
+  return snprintf(dst, dst_size, "%ld.%06lu", tv.tv_sec, tv.tv_usec);
+#endif
+}
+
+static int _w_vprintf(const char *fmt, va_list ap)
+{
+  char fmt2[FMT_BUF_SIZE];
+  int fmt2_len;
+  int ret;
+
+  fmt2_len = _w_now(fmt2, sizeof(fmt2));
+  if (strlen(fmt) > (sizeof(fmt2) - fmt2_len - 2 /* , & NUL */)) {
+    /* Our static buffer isn't big enough, so call vprintf() as is. */
+    ret = vprintf(fmt, ap);
+  } else {
+    fmt2[fmt2_len] = ',';
+    memcpy(fmt2 + fmt2_len + 1, fmt, strlen(fmt) + 1);
+    ret = vprintf(fmt2, ap);
+  }
+  va_end(ap);
+  return ret;
+}
+
+static void _w_initialize_labels()
+{
+  int i;
+  char buf[16];
+
+  for (i = 0; i < MAX_SEVERITY+1; i++) {
+    snprintf(buf, sizeof(buf), "sev-%d", i);
+    _severity_labels[i] = strdup(buf);
+  }
+  for (i = 0; i < MAX_CATEGORY+1; i++) {
+    snprintf(buf, sizeof(buf), "cat-%d", i);
+    _category_labels[i] = strdup(buf);
+    _cat2sev_threshold[i] = MAX_SEVERITY;
+  }
+  _labels_initialized = 1;
+}
+
+static int _w_l(unsigned char severity, unsigned char category,
+  const char *fmt, va_list ap)
+{
+  char fmt2[FMT_BUF_SIZE];
+
+  if (severity > MAX_SEVERITY || category > MAX_CATEGORY ||
+      severity > _cat2sev_threshold[category]) {
+    return 0;
+  }
+  if (! _labels_initialized) {
+    _w_initialize_labels();
+  }
+
+  snprintf(fmt2, sizeof(fmt2), "%s,%s,%s\n",
+    _severity_labels[severity], _category_labels[category], fmt);
+  return _w_vprintf(fmt2, ap);
+}
+
+/********************/
+/* Public functions */
+/********************/
+
+int printf(const char *fmt, ...)
```

Comment:
What is this function doing? Is this leftover from the initial draft?

Reply:
It's there to provide timestamps on printf messages until conversion away from printf is complete, TBD.